src/Native/Unix/System.IO.Ports.Native/pal_serial.c

```diff
@@ -6,11 +6,27 @@
 #include <fcntl.h>
 #include <errno.h>
 #include <pal_serial.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
 
 /* Open device file in non-blocking mode and without controlling terminal */
 intptr_t SystemIoPortsNative_SerialPortOpen(const char * name)
 {
-    intptr_t result;
-    while ((result = open(name, O_RDWR | O_NOCTTY | O_CLOEXEC | O_NONBLOCK)) < 0 && errno == EINTR);
-    return result;
+    intptr_t fd;
+    while ((fd = open(name, O_RDWR | O_NOCTTY | O_CLOEXEC | O_NONBLOCK)) < 0 && errno == EINTR);
+
+    if (fd < 0)
+    {
+        return fd;
+    }
+
+    if (ioctl(fd, TIOCEXCL) != 0)
+    {
+        // We couldn't get exclusive access to the device file
+        close(fd);
```

Comment:
We should save off the error information and restore it so that close doesn't impact it, e.g.
```C++
int oldErrno = errno;
close(fd);
errno = oldErrno;
```

Reply:
good point